### ISSUE TITLE
PortAudio + Fork + PulseAudio

### ISSUE THREAD

**Issue by dubvulture** [PortAudio/portaudio#964]
**Describe the bug**
If `Pa_Initialize` is already called by a parent process, PortAudio does not report the `pulse` device when called from any forked child process.
If the parent process does not call `Pa_Initialize`, then any child process will find `pulse` but the total number of available devices changes.

**To Reproduce**
```cpp
#include <iostream>
#include <unistd.h>

#include <portaudio.h>

void portaudio_test(int child_num)
{
   Pa_Initialize();
   PaDeviceIndex count = Pa_GetDeviceCount();

   int pulse_idx = -1;
   for (int i = 0; i < count; ++i) {
      auto deviceInfo = Pa_GetDeviceInfo(i);
      auto name = std::string(deviceInfo->name).substr(0, 5);
      if (name == "pulse") {
         pulse_idx = i;
         break;
      }
   }
   if (child_num == -1) {
      std::cout << "Parent "; 
   } else {
      std::cout << "Child#" << child_num;
   }
   std::cout << " => pulse idx: " << pulse_idx << ", count = " << count << "\n";
   Pa_Terminate();
}

int main()
{
   // comment out to disable in parent process
   portaudio_test(-1);

   for (int i = 0; i < 4; ++i) {
      auto pid = fork();
      if (pid < 0) {
         // failed to fork
         exit(1);
      }
      else if (pid == 0) {
         // child process
         portaudio_test(i);
         exit(0);
      } else {
         // parent process, just iterate anf fork again
      }
   }
}
```
Compiled with gcc 11.4:
```bash
g++ main.cpp -lportaudio
```

**Expected behavior**
```
Parent   => pulse idx: 10, count = 14
Child#0  => pulse idx: 10, count = 14
Child#1  => pulse idx: 10, count = 14
Child#2  => pulse idx: 10, count = 14
Child#3  => pulse idx: 10, count = 14
```
The order of the lines and the actual `indeces` / `count` may vary by machine, but they should be consistent across processes.

**Actual behavior**
If both parent and child processes run `portaudio_test`:
```
Parent  => pulse idx: 10, count = 14
Child#1 => pulse idx: -1, count = 8
Child#0 => pulse idx: -1, count = 7
Child#3 => pulse idx: -1, count = 9
Child#2 => pulse idx: -1, count = 10
```
If I disable the `portaudio_test` call from the parent (first line in main):
```
Child#3 => pulse idx: 5, count = 9
Child#2 => pulse idx: 5, count = 9
Child#0 => pulse idx: 4, count = 8
Child#1 => pulse idx: 8, count = 12
```

**Desktop:**
 - OS: Ubuntu 22.04
 - PortAudio version: 19.6.0-1.1 (from Ubuntu repositories)

**Comment by RossBencina** [comment]
What happens if you call `portaudio_test` twice from the parent prior to forking? Does it work correctly both times? How about in the parent after forking?

**Comment by dubvulture** [comment]
Twice before forking:
```
Parent  => pulse idx: 10, count = 14
Parent  => pulse idx: 10, count = 14
Child#2 => pulse idx: -1, count = 10
Child#1 => pulse idx: -1, count = 9
Child#0 => pulse idx: -1, count = 11
Child#3 => pulse idx: -1, count = 12
```

Before and after forking (without waiting for any child process):
```
Parent  => pulse idx: 10, count = 14
Child#2 => pulse idx: -1, count = 9
Child#0 => pulse idx: -1, count = 9
Child#3 => pulse idx: -1, count = 9
Child#1 => pulse idx: -1, count = 11
Parent  => pulse idx: 5, count = 9
```
After forking, the parent count changes on each run, but the pulse index is always non-negative.

**But** if I explicitly wait for all forked process to terminate, the count remains correct:
```
Parent  => pulse idx: 10, count = 14
Child#0 => pulse idx: -1, count = 12
Child#3 => pulse idx: -1, count = 9
Child#2 => pulse idx: -1, count = 10
Child#1 => pulse idx: -1, count = 11
Parent  => pulse idx: 10, count = 14
```

Another test, if I do **not** call `Pa_Terminate` in the parent process, the output is the expected one for **any** of the previous combinations of tests:
```
Parent  => pulse idx: 10, count = 14
Child#0 => pulse idx: 10, count = 14
Child#1 => pulse idx: 10, count = 14
Child#2 => pulse idx: 10, count = 14
Child#3 => pulse idx: 10, count = 14
```

**Comment by illuusio** [comment]
Thank you for reporting this. Just to make sure that you are using Portaudio from Ubuntu not compiled from git. As (correct me if I'm wrong) Ubuntu 22.04 still uses Pulseaudio gateway from Alsa which is bit obscure to debug. It was reason why I write hostapi in first place. But if this is what happens in git code also then I can help to debug. Just for curiosity when do you need forking in audio code?

**Comment by dubvulture** [comment]
Yes, I was using portaudio from the Ubuntu repositories.
I don't really know much about the audio stack on Linux, but I can tell for sure that I'm using `pipewire-media-session-pulseaudio`, so `pactl info` reports the server as `PulseAudio (on PipeWire 0.3.48)`, which now introduces more variables I guess.

So before typing this I spun up two fresh VMs:
- Debian 11 (`pactl info` reports `pulseaudio`)
- Debian 12 (`pactl info` reports `PulseAudio (on PipeWire 0.3.65)`)

And I was able to reproduce the problem on both platforms.
Now I guess I can try compiling and using portaudio from git and see what happens.

Regarding the forking use case, well that's not really on me, I was just trying to reproduce an issue I had while using `pyaudio` in Python, which expects the following workflow:
```python
 p = pyaudio.PyAudio() # Pa_Initialize
 ...
 p.terminate() # Pa_Terminate
```
And I was using all of this inside the main python process and across multiple `multiprocessing.Process` (which basically forks the interpreter as in my example).

**Comment by dubvulture** [comment]
### Debian 11 / Debian 12 (no difference in behaviour between the two)
#### master
```
Configuration summary:

  Target ...................... x86_64-unknown-linux-gnu
  C++ bindings ................ no
  Debug output ................ no

  ALSA ........................ yes
  ASIHPI ...................... no

  AudioIO ..................... no
  OSS ......................... yes
  JACK ........................ no
  PulseAudio .................. no
  Sndio ....................... no
```
```
Parent  => pulse idx: 4, count = 9
Child#2 => pulse idx: -1, count = 5
Child#3 => pulse idx: -1, count = 6
Child#0 => pulse idx: -1, count = 7
Child#1 => pulse idx: -1, count = 7
```

#### v19.7.0
```
Configuration summary:

  Target ...................... x86_64-unknown-linux-gnu
  C++ bindings ................ no
  Debug output ................ no

  ALSA ........................ yes
  ASIHPI ...................... no

  OSS ......................... yes
  JACK ........................ no
```
```
Parent  => pulse idx: 11, count = 16
Child#1 => pulse idx: -1, count = 6
Child#3 => pulse idx: -1, count = 9
Child#0 => pulse idx: -1, count = 12
Child#2 => pulse idx: -1, count = 10
```

**Comment by dubvulture** [comment]
Ok, I recompiled portaudio from master with `PA_ENABLE_DEBUG_OUTPUT` and...
Parent:
```
FillInDevInfo: Filling device info for: pulse
GropeDevice: collecting info..
GropeDevice: collecting info..
FillInDevInfo: Adding device pulse: 4
```
Child:
```
FillInDevInfo: Filling device info for: pulse
OpenPcm: Opened device 'pulse' ptr[(nil)] - result: [-5:Input/output error]
OpenPcm: Opened device 'pulse' ptr[(nil)] - result: [-5:Input/output error]
FillInDevInfo: Skipped device: pulse, all channels == 0
```

So the failure is [here](https://github.com/PortAudio/portaudio/blob/57aa393109ec996799d3a5846c9ecb0a65b64644/src/hostapi/alsa/pa_linux_alsa.c#L1148) and maybe it's a problem with alsa pulse plugin itself?

**Comment by dubvulture** [comment]
Ok, the problem is pulseaudio itself, it keeps track of whether the running process has been forked and bails if it detects so.
This [`pa_detect_fork`](https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/blob/master/src/pulse/fork-detect.c#L50) is used extensively in all the `pa_context_*` functions and in this case it prevents `pa_context_new` from completing successfully when portaudio/alsa/whatever tries to set it up in a forked process, even if the context was correctly closed.
I tried recompiling pulseaudio while disabling this check and rerunning the previous tests, and the children now correctly initialize the alsa/pulse plugin.

This explains most of the previously observed behaviours, even:
> Another test, if I do not call Pa_Terminate in the parent process, the output is the expected one for any of the previous combinations of tests:
> ```
> Parent  => pulse idx: 10, count = 14
> Child#0 => pulse idx: 10, count = 14
> Child#1 => pulse idx: 10, count = 14
> Child#2 => pulse idx: 10, count = 14
> Child#3 => pulse idx: 10, count = 14
> ```
since any subsequent `Pa_Initialize` just returns immediately and never tries to initialize alsa/pulse.

I'm still not sure on why the device count varies between each child process but there might be something similar going on with other plugins.

Closing and moving this to the pulseaudio issue tracker.
